src/base/utils/tar.h

```diff
@@ -0,0 +1,112 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <QByteArray>
+#include <QString>
+
+namespace Utils
+{
+    namespace Tar
+    {
+        QHash<QString, QByteArray> untar(const QByteArray &data, QString &errorString);
+
+        class Header
+        {
+        public:
+            enum class Types : unsigned char
+            {
+                REG_FILE_NULL = '\0',
+                REG_FILE_ASCII = '0',
+                // We don't care about the other file types
+
+                UNKNOWN = std::numeric_limits<unsigned char>::max()
+            };
+
+            enum class State
+            {
+                EMPTY,
+                INVALID,
+                VALID
+            };
+
+            explicit Header(const QByteArray &header);
+            Header() = default;
+
+            QString fileName() const;
+            Types fileType() const;
+            int fileSize() const;
+
+            State state() const { return m_state; }
+
+        private:
+            int octalToDec(const QByteArray &data) const;
+
+            // The various header fields that hold strings are of known max size.
+            // So in the case the string occupies their max size, the string will not
+            // be null terminated. It is pointless to have std::strlen() run through
+            // the whole buffer when we already know the upper limit. Also the buffer
+            // isn't guaranteed to contain a null byte and std::strlen() will have
+            // undefined behavior.
+            int strlen(const QByteArray &str) const { return qstrnlen(str.constData(), str.length()); }
+
+            // Verified: When the checksum field gets verified
+            // Null: When all bytes are null. Used to indicate end-of-archive (EOF).
+            // Since both actions require running through the whole buffer, combine them to avoid running
+            // twice through the buffer.
+            void verifyAndCheckNull();
+
+            QByteArray m_header;
+            State m_state = State::INVALID;
+        };
+
+        class StreamReader
+        {
+        public:
+            explicit StreamReader(const QByteArray &data);
+            explicit StreamReader(const QByteArray &&data) = delete;
+
+            Header::Types readNext();
```

Comment:
Why don't you go the common way? Just return the status of successful reading as a bool. Other information is meaningless at this point.